kem/src/kem.rs

```diff
@@ -1,48 +1,81 @@
 //! KEM Traits
 
 use crate::errors::Error;
+
 use core::fmt::Debug;
-use generic_array::{ArrayLength, GenericArray};
 
+use generic_array::{ArrayLength, GenericArray};
 use rand_core::{CryptoRng, RngCore};
+use zeroize::{Zeroize, ZeroizeOnDrop};
 
 /// Trait impl'd by concrete types that represent an encapsulated key. This is intended to be, in
 /// essence, a bag of bytes.
 pub trait EncappedKey: AsRef<[u8]> + Debug + Sized {
-    /// The size of the shared secret that this KEM produces
-    type NSecret: ArrayLength<u8>;
+    /// The size, in bytes, of an encapsulated key.
+    type EncappedKeySize: ArrayLength<u8>;
+
+    /// The size, in bytes, of the shared secret that this KEM produces.
+    type SharedSecretSize: ArrayLength<u8>;
 
     /// Represents the identity key of an encapsulator. This is used in authenticated
     /// decapsulation.
     type SenderPublicKey;
 
     /// The public key of a decapsulator. This is used in encapsulation.
     type RecipientPublicKey;
+
+    /// Parses an encapsulated key from its byte representation.
+    fn from_bytes(bytes: &GenericArray<u8, Self::EncappedKeySize>) -> Result<Self, Error>;
+
+    /// Borrows a byte slice representing the serialized form of this encapsulated key.
+    fn as_bytes(&self) -> &GenericArray<u8, Self::EncappedKeySize> {
+        // EncappedKey is already AsRef<[u8]>, so we don't need to do any work. This will panic iff
+        // the underlying bytestring is not precisely NEnc bytes long.
+        self.as_ref().into()
+    }
+}
+
+/// The shared secret that results from key exchange.
+#[derive(Zeroize)]
+pub struct SharedSecret<EK: EncappedKey>(GenericArray<u8, EK::SharedSecretSize>);
+
+impl<EK: EncappedKey> ZeroizeOnDrop for SharedSecret<EK> {}
```

Comment:
I think you likely want this:

```suggestion
#[derive(ZeroizeOnDrop)]
pub struct SharedSecret<EK: EncappedKey>(GenericArray<u8, EK::SharedSecretSize>);
```

What you have doesn't add a `Drop` handler, it just impls the `Zeroize` trait. But it's also a case where the *only* thing that should call `zeroize` is the `Drop` impl (i.e. you don't want to zeroize a shared secret you intend to still use), which is what the `ZeroizeOnDrop` custom derive is useful for.

Also, this is a simple enough case you can write it by hand without using the custom derive support:

```rust
impl<EK: EncappedKey> Drop for SharedSecret<EK> {
    fn drop(&mut self) {
        self.0.zeroize();
    }
}

impl<EK: EncappedKey> ZeroizeOnDrop for SharedSecret<EK> {}
```

Reply:
Good catch! Just pushed the fix.